src/integration-test/java/com/commercetools/sync/integration/externalsource/cartdiscounts/CartDiscountSyncIT.java

```diff
@@ -0,0 +1,643 @@
+    }
+
+    @Test
+    void sync_WithNullDraft_ShouldExecuteCallbackOnErrorAndIncreaseFailedCounter() {
+        //preparation
+        final CartDiscountDraft newCartDiscountDraft = null;
+
+        final List<String> errorMessages = new ArrayList<>();
+        final List<Throwable> exceptions = new ArrayList<>();
+
+        final CartDiscountSyncOptions cartDiscountSyncOptions = CartDiscountSyncOptionsBuilder
+            .of(CTP_TARGET_CLIENT)
+            .errorCallback((errorMessage, exception) -> {
+                errorMessages.add(errorMessage);
+                exceptions.add(exception);
+            })
+            .build();
+
+        final CartDiscountSync cartDiscountSync = new CartDiscountSync(cartDiscountSyncOptions);
+
+        //test
+        final CartDiscountSyncStatistics cartDiscountSyncStatistics = cartDiscountSync
+            .sync(singletonList(newCartDiscountDraft))
+            .toCompletableFuture()
+            .join();
+
+        //assertions
+        assertThat(errorMessages)
+            .hasSize(1)
+            .hasOnlyOneElementSatisfying(message ->
+                assertThat(message).isEqualTo("Failed to process null cart discount draft.")
+            );
+
+        assertThat(exceptions)
+            .hasSize(1)
+            .hasOnlyOneElementSatisfying(throwable -> assertThat(throwable).isNull());
+
+        assertThat(cartDiscountSyncStatistics).hasValues(1, 0, 0, 1);
+    }
+
+    @Test
+    void sync_WithSeveralBatches_ShouldReturnProperStatistics() {
+        // preparation
+        final List<String> sortOrders = getSortOrders(100);
+        // Default batch size is 50 (check CartDiscountSyncOptionsBuilder) so we have 2 batches of 50
+        final List<CartDiscountDraft> cartDiscountDrafts = IntStream
+            .range(0, 100)
+            .mapToObj(i ->
+                CartDiscountDraftBuilder.of(LocalizedString.of(Locale.ENGLISH, "KEY_" + Integer.toString(i)),
+                    CART_DISCOUNT_CART_PREDICATE_2,
+                    CART_DISCOUNT_VALUE_2,
+                    CART_DISCOUNT_TARGET_2,
+                    sortOrders.get(i),
+                    false)
+                                        .active(false)
+                                        .build())
+            .collect(Collectors.toList());
+
+        final CartDiscountSyncOptions cartDiscountSyncOptions = CartDiscountSyncOptionsBuilder
+            .of(CTP_TARGET_CLIENT)
+            .build();
+
+        final CartDiscountSync cartDiscountSync = new CartDiscountSync(cartDiscountSyncOptions);
+
+        //test
+        final CartDiscountSyncStatistics cartDiscountSyncStatistics = cartDiscountSync
+            .sync(cartDiscountDrafts)
+            .toCompletableFuture()
+            .join();
+
+
+        //assertion
+        assertThat(cartDiscountSyncStatistics).hasValues(100, 100, 0, 0);
+    }
+
+    @Nonnull
+    private SphereClient buildClientWithConcurrentModificationUpdate() {
```

Comment:
please put the method declaration after it has been called.

Reply:
94322af